amaranth_soc/gpio.py

```diff
@@ -0,0 +1,333 @@
+from amaranth import *
+from amaranth.lib import enum, wiring
+from amaranth.lib.wiring import In, Out, flipped, connect
+
+from . import csr
+
+
+__all__ = ["PinMode", "PinSignature", "Peripheral"]
+
+
+class PinMode(enum.Enum, shape=unsigned(2)):
+    """GPIO pin mode.
+
+    The 2-bit values of this enumeration can be written to a :class:`Peripheral.Mode` field to
+    configure the pins of a :class:`Peripheral`.
+    """
+
+    #: `Input-only` mode.
+    #:
+    #: The pin output is disabled but remains connected to its :class:`Peripheral.Output` field.
+    #: Its :attr:`Peripheral.alt_mode` bit is wired to 0.
+    INPUT_ONLY = 0b00
+
+    #: `Push-pull` mode.
+    #:
+    #: The pin output is enabled and connected to its :class:`Peripheral.Output` field. Its
+    #: :attr:`Peripheral.alt_mode` bit is wired to 0.
+    PUSH_PULL = 0b01
+
+    #: `Open-drain` mode.
+    #:
+    #: The pin output is enabled when the value of its :class:`Peripheral.Output` field is 0, and
+    #: is itself wired to 0. Its :attr:`Peripheral.alt_mode` bit is wired to 0.
+    OPEN_DRAIN = 0b10
+
+    #: `Alternate` mode.
+    #:
+    #: The pin output is disabled but remains connected to its :class:`Peripheral.Output` field.
+    #: Its :attr:`Peripheral.alt_mode` bit is wired to 1.
+    ALTERNATE = 0b11
+
+
+class PinSignature(wiring.Signature):
+    """GPIO pin signature.
+
+    Interface attributes
+    --------------------
+    i : :class:`Signal`
+        Input.
+    o : :class:`Signal`
+        Output.
+    oe : :class:`Signal`
+        Output enable.
+    """
+    def __init__(self):
+        super().__init__({
+            "i":  In(unsigned(1)),
+            "o":  Out(unsigned(1)),
+            "oe": Out(unsigned(1)),
+        })
+
+
+class Peripheral(wiring.Component):
+    class Mode(csr.Register, access="rw"):
+        """Mode register.
+
+        This :class:`csr.Register` contains an array of ``pin_count`` read/write fields. Each field
+        is 2-bit wide and its possible values are defined by the :class:`PinMode` enumeration.
+
+        If ``pin_count`` is 4, then the register has the following fields:
```

Comment:
Should we use 8 as an example pin count? It's very rare to see GPIO banks of 4 pins, and it feels better to have something common on display.

Reply:
Agreed, changed to 8 pins.